Feature/sim 1151/rewrite variability mixin

danielsf opened the pull request:
This pull request creates a list _all_available_columns which contains the names of all of the columns a given InstanceCatalog knows how to calculate.  This is how the new Variability mixins can tell whether or not they contain the delta_magnitude columns necessary for implementing variability.

danielsf committed:
added member variable self._all_getters to InstanceCatalog

This is a list containing the names of all of the columns
for which getter methods exist

danielsf committed:
made db generation methods in testColumnOrigins.py more
re-usable

danielsf committed:
added unit test for _all_getters

danielsf committed:
test for columns that depend on each other only if they are defined

danielsf committed:
replace _all_getters with _all_available_columns so that we
can even consider columns defined in the database

danielsf committed:
now _all_available_columns will also track the names of columns
set by default

danielsf committed:
expanded the comment explaining _all_available_columns

danielsf committed:
renamed unit test class for all_available_columns

danielsf committed:
renamed self.all_calculated_columns to self._actually_calculated_columns
to avoid confusion with self._all_available_columns

rbiswas4 commented on python/lsst/sims/catalogs/measures/instance/InstanceCatalog.py:
```diff
@@ -273,6 +273,35 @@ def __init__(self, db_obj, obs_metadata=None, column_outputs=None,
+        #is called in case any getters depend on the contents of
+        #_all_available_columns.  That way, self._check_requirements()
+        #can verify that the getter will run the way it is actually
+        #being called.
```
spaces after # (if we are following pep8)

rbiswas4 commented on python/lsst/sims/catalogs/measures/instance/InstanceCatalog.py:
```diff
@@ -273,6 +273,35 @@ def __init__(self, db_obj, obs_metadata=None, column_outputs=None,
+                columnName = name[8:]
+                if columnName not in self._all_available_columns:
+                    self._all_available_columns.append(columnName)
+
```
I am not sure if this makes all items in the list self._all_available_columns unique? Could items from columnMap.keys(), default_cols also be used in _compound_column_names? D

Does this matter?

danielsf replied:
self.column_by_name has a specific resolution order so that it is possible for a getter to override a database value to override a default value.

I just realized that what I did is somewhat redundant with the method _all_columns(), except that _all_columns does not correctly break down compound columns into their constituent sub-columns.  Let me deal with that and push some changes.

danielsf committed:
removed redundant _all_columns() method from InstanceCatalog

Note that _all_columns() did not perform as promised.  The strip()
method did not remove the specified words; it removed all leading instances
of the letters in those words, i.e.

'default_formats'.strip('default_') =  'ormats'

because the strip() removed all leading 'f's

danielsf committed:
added safety check when setting default _columns_outputs

rbiswas4 commented on tests/testColumnOrigins.py:
```diff
@@ -6,11 +6,11 @@
 from lsst.sims.catalogs.generation.db import CatalogDBObject
 
-def makeTestDB(size=10, **kwargs):
+def makeTestDB(size=10, name=None, **kwargs):
```
the default argument for name will not run, I think? so, why not make it mandatory and change the order? or give it a default name that will actually run?

rbiswas4 commented on tests/testColumnOrigins.py:
```diff
@@ -6,11 +6,11 @@
-def makeTestDB(size=10, **kwargs):
+def makeTestDB(size=10, name=None, **kwargs):
     """
     Make a test database
```
Since this is not a 'private' function, should arguments be explained in the doc strings?

danielsf committed:
reformatted comments in InstanceCatalog.py so that spaces come after '#'

danielsf committed:
reformatted makeTestDB in testColumnsOrigins.py so that you
must pass in a name for the database file

danielsf committed:
had to change how testColumnOrigins.py refers to its test database
in order to be self-consistent with the new database access API
